syntax = "proto3";

package sentinel.v1;

option cc_enable_arenas = true;
option java_multiple_files = true;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

// Envelope for any observed kernel-level event.
message EventEnvelope {
  // Unique ID for correlating responses.
  string event_id = 1;
  // Node and workload metadata.
  string hostname = 2;
  string pod_name = 3;
  string namespace = 4;
  string container_id = 5;
  // Generic event name and ordered data vector (strings) for efficiency.
  // Canonical vector order for curated syscall events:
  // [event_name, event_id, comm, pid, tid, uid, arg0, arg1, path, flags]
  // Missing values are encoded as empty string.
  // - arg0/arg1 carry event-specific numeric arguments when present.
  // - path is used by path-based events (open/exec/unlink/rename/chmod/chown).
  // - flags is used by open/openat/openat2.
  string event_type = 9;
  repeated string data = 10;
  // Additional labels or enrichment.
  map<string, string> attributes = 20;
  // Kernel timestamp (nanoseconds).
  uint64 ts_unix_nano = 30;
}

message DetectionResponse {
  string event_id = 1;
  bool anomaly = 2;
  string reason = 3;
  double score = 4;
  google.protobuf.Timestamp ts = 5;
}

message AnomalyReport {
  string event_id = 1;
  string reason = 2;
  double score = 3;
  map<string, string> labels = 4;
}

service DetectorService {
  // Bidirectional stream of events and anomaly responses.
  rpc StreamEvents (stream EventEnvelope) returns (stream DetectionResponse);
  // Optional explicit anomaly report push.
  rpc ReportAnomaly (AnomalyReport) returns (google.protobuf.Empty);
}
